Add Makefile, configuration files and scripts

StefanJum opened the pull request:
Update the README with instructions on setting up, configuring, building and running the helloworld app with Unikraft. As part of it, add Makefile files, configuration files and scripts; they are referenced in the `README.md` file.

Current instructions cover the use of QEMU/KVM platform (both on x86_64 and on AArch64) and GCC. Firecraker/KVM, Xen, Linux platforms and Clang are not yet documented.

StefanJum committed:
Add Makefile, configuration files and scripts

Update the README with instructions on setting up, configuring, building
and running the helloworld app with Unikraft. As part of it, add Makefile
files, configuration files and scripts; they are referenced in the
`README.md` file.

Current instructions cover the use of QEMU/KVM platform (both on x86_64 and
on AArch64) and GCC. Firecraker/KVM, Xen, Linux platforms and Clang are
not yet documented.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand commented on run-qemu-aarch64:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
```suggestion
#!/bin/sh
```

razvand commented on run-qemu-x86_64:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
```suggestion
#!/bin/sh
```

razvand commented on README.md:
````diff
@@ -1,63 +1,261 @@
+     $ git clone https://github.com/unikraft/unikraft unikraft
+     ```
+
+  1. While inside the `.unikraft/` directory, create the `libs/` directory:
+
+     ```console
+     $ mkdir libs
+     ```
+
````
Remove this. No `libs/` directory is required.

razvand commented on README.md:
````diff
@@ -1,63 +1,261 @@
+     ```console
+     $ tree -F -L 2 .unikraft/
+     .unikraft/
+     |-- libs/
````
Remove `libs/`.

razvand commented on README.md:
```diff
@@ -1,63 +1,261 @@
+* `git`
+* `wget`
+* `uuid-runtime`
+* `qemu-system-x86`
```
```suggestion
* `qemu-system-x86`
* `qemu-system-arm`
```

razvand commented on README.md:
```diff
@@ -1,63 +1,261 @@
+  wget \
+  uuid-runtime \
+  qemu-kvm \
+  qemu-system-x86 \
```
```suggestion
  qemu-system-x86 \
  qemu-system-arm \
```

razvand commented on README.md:
````diff
@@ -1,63 +1,261 @@
-```
-$ make menuconfig
+```console
+$ apt-get install -y --no-install-recommends \
````
```suggestion
$ sudo apt install -y --no-install-recommends \
```

razvand reviewed (changes requested):
Thanks, @StefanJum. See my comments.

StefanJum commented:
Done @razvand

razvand commented on Makefile:
```diff
@@ -1,5 +1,5 @@
+UK_LIBS ?= $(PWD)/.unikraft/libs
 LIBS :=
 
 all:
```
Why are there some `Makefile` lines missing? Such as those here: https://github.com/unikraft/app-nginx/blob/staging/Makefile

StefanJum replied:
What do you mean missing? The `Makefile` was already in the repository so it's just updated, it's still all [there](https://github.com/unikraft/app-helloworld/blob/5a2b7dbec0f478a4c905d55b503ab5933158cbe8/Makefile).

razvand replied:
@StefanJum , you're right. I'm stupid.

razvand reviewed (changes requested):
Add a `.gitignore` file. Probably as a new commit.

StefanJum committed:
Add .gitignore file

Ignore files and folders that are output from the set up, configure,
build or run steps.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum commented:
@razvand added the `.gitignore`, see the comment about the `Makefile`.

razvand reviewed (approved):
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

nderjung reviewed (changes requested):
In my opinion, we should not be adding `.config*` files to the repository additional scripts on the top-level (at least, in the state that they are in: they should have relative data (e.g. user paths) and `unset` attributes removed.) Additionally, the tutorial contains the same information as the scripts, these should be removed. Finally, and most importantly, `kraft` was entirely removed from the README. In my opinion this is a mistake. We are aiming to make [KraftKit](https://kraftkit.sh) the first-class and go-to-choice for using and building Unikraft. The README should start with this, and progress with "for advanced, or first-principle usage"

StefanJum commented:
> In my opinion, we should not be adding `.config*` files to the repository additional scripts on the top-level (at least, in the state that they are in: they should have metadata and `unset` attributes removed. Additionally, the tutorial contains the same information as the scripts, these should be removed.

What `scripts`?

> Finally, and most importantly, `kraft` was entirely removed from the README. In my opinion this is a mistake. We are aiming to make [KraftKit](https://kraftkit.sh) the first-class and go-to-choice for using and building Unikraft. The README should start with this, and progress with "for advanced, or first-principle usage"

Yes, we can do that after `kraftkit` documentation is in order. The main (short term) goal for the changes is to get very new people to quickly setup the application, mostly as part of USoC (which will happen tomorrow, 04 July, not in time for `kraftkit` to be properly tested and documented).

If people will see `kraftkit` as the top choice they will try using it, and in the current state they will very likely fail, which will get them frustrated and cause them to give up, so we want the instructions to be 100% working first, then when we are ready we can have the instructions for using `kraftkit` as very strongly recommended (we can even remove all the running scripts / config files at that point).

nderjung commented:
Hi @StefanJum,

> Which scripts?

The new scripts `run-qemu-aarch64` and `run-qemu-x86_64`.

> Yes, we can do that after kraftkit documentation is in order. The main (short term) goal for the changes is to get very new people to quickly setup the application, mostly as part of USoC (which will happen tomorrow, 04 July, not in time for kraftkit to be properly tested and documented).

> If people will see kraftkit as the top choice they will try using it, and in the current state they will very likely fail, which will get them frustrated and cause them to give up, so we want the instructions to be 100% working first, then when we are ready we can have the instructions for using kraftkit as very strongly recommended (we can even remove all the running scripts / config files at that point).

No. We must direct people to KraftKit regardless of the state of the documentation (which I am working on full-time now). Thi README is part of the same documentation and waiting for one piece elsewhere is not the right approach. We should update all relevant documents to usage with KraftKit. Finally, there are admittedly bugs in KraftKit, like there are with Unikraft itself. This document only serves one persona and does not help us in getting valuable bug reports and user experience stories with KraftKit. Redirecting people away will not benefit us, short-, mid- or long-term.

Finally, the relevant USoC channels should contain information about how to get started, in the USoC website and on Discord, outside of this README. This README should not be tailored for USoC.

StefanJum commented:
> The new scripts `run-qemu-aarch64` and `run-qemu-x86_64`.

Then what do you mean by `the tutorial contains the same information as the scripts`, the scripts have just the commands used for running, nothing more, which is not explained in the tutorial, just referenced as in `run this script to start the app`.

> No. We must direct people to KraftKit regardless of the state of the documentation (which I am working on full-time now). Thi README is part of the same documentation and waiting for one piece elsewhere is not the right approach. We should update all relevant documents to usage with KraftKit. Finally, there are admittedly bugs in KraftKit, like there are with Unikraft itself. This document only serves one persona and does not help us in getting valuable bug reports and user experience stories with KraftKit. Redirecting people away will not benefit us, short-, mid- or long-term.

I highly doubt that newcomers will start using a new project / tool / whatever (being that `Unikraft` or `Kraftkit`), bump into issues and then report them. The very big majority of people will quit trying after 10 minutes of running into issues, the reporting happens after they see the project working, then run into issues while exploring. That is why the first instructions they see should always be working.

`direct people to KraftKit` should mean sending people somewhere where they can find detailed instructions on how to use `kraftkit`, or at least giving actual instructions on how to run the app using `kraftkit`, which right now did not work out of the box for me. I would **really** not put something that does not work outside the box for everyone in the README of `app-helloworld`.

razvand commented:
@nderjung, this is easier for everyone. Once KraftKit comes about, we'll just update the README.

We can also make plenty of updates such as moving the documentation for **each** app in centralized repository, and provide extensive instructions and improve things. But that's not immediate. We simply make it work now and update it later. Good now is better than perfect sometime later. We can make it perfect at a later time. As long as nothing breaks (and it doesn't) there is no problem.

nderjung commented on README.md:
```diff
@@ -1,63 +1,256 @@
-# Unikraft "hello world" Application
+# Helloworld on Unikraft
 
-This application prints a basic "Hello World!" message.
+This application starts a Helloworld application with Unikraft.
+Follow the instructions below to set up, configure, build and run Helloworld.
 
-To configure, build and run the application you need to have [kraft](https://github.com/unikraft/kraft) installed.
+## Requirements
```
# A simple "Hello, world" application on Unikraft

To get started immediately, you can use [Unikraft's companion command-line companion tool `kraft`](https://kraftkit.sh). Simply start by running the interactive installer:
```
curl --proto '=https' --tlsv1.2 -sSf https://get.kraftkit.sh | sh
```

Once installed, simple clone this repository and run `kraft build`:
```
git clone --depth=1 https://github.com/unikraft/app-hellworld.git
cd app-helloworld;
kraft build
```

This will guide you through an interactive build process where you can select one of the available targets (architecture/platform combinations). Otherwise, we recommend building for qemu/x86_64 like so:
```
kraft build --arch x86_64 --plat qemu
```

Once built, you can instantiate the unikernel via:
```
kraft run
```

When left without any input flags, you'll be queried for the desired target architecture/platform.

## Build Unikraft from first principles (Advanced)

nderjung reviewed (commented):
Please include the preamble (see below) at the start of the README and:

* Remove the `run-qemu-aarch64` and `run-qemu-x86_64` scripts, adjust the README to include what they include. No need to dirty the repo with scripts that contain 1-liners;
* Remove all `# XXX is not set` config options;
* Remove the attributes `CONFIG_UK_BASE` and `CONFIG_UK_APP` as they contain your personal directories and they are overwritten by Unikraft's build system;
* Remove `CONFIG_UK_FULLVERSION` and `CONFIG_UK_CODENAME` options, these are also overwritten;
* Rename `config-qemu-aarch64` to `.config.helloworld_qemu-arm64`
* Rename `config-qemu-x86_64` to `.config.helloworld_qemu-x86_64`

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
-# Unikraft "hello world" Application
+# A simple "Hello, world!" application on Unikraft
```
```suggestion
# A Simple "Hello, world!" Application on Unikraft
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
 
-To configure, build and run the application you need to have [kraft](https://github.com/unikraft/kraft) installed.
+To get started immediately, you can use Unikraft's companion command-line companion tool, [`kraft`](https://github.com/unikraft/kraftkit).
+Simply start by running the interactive installer:
```
```suggestion
Start by running the interactive installer:
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
-And, finally, run the application:
+When left without any input flags, you'll be queried for the desired target architecture/platform.
+
+## Build Unikraft from first principles (Advanced)
```
```suggestion
## Work with the Basic Build & Run Toolchain (Advanced)
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+
+## Build Unikraft from first principles (Advanced)
+
+You can build the application from ground, without using the companion tool `kraft`.
```
```suggestion
You can set up, configure, build and run the application from grounds up, without using the companion tool `kraft`.
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+git clone https://github.com/unikraft/app-helloworld helloworld
+cd helloworld/
+mkdir .unikraft
+git clone https://github.com/unikraft/unikraft .unikraft/unikraft
+UK_DEFCONFIG=$(pwd)/.config.helloworld_qemu-x86_64 make defconfig
```
```suggestion
git clone https://github.com/unikraft/unikraft .unikraft/unikraft
cd ..
UK_DEFCONFIG=$(pwd)/.config.helloworld_qemu-x86_64 make defconfig
```

StefanJum replied:
I'm not entering the `.unikraft` directory. Why add `cd ..`?

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
-$ kraft run
-Hello World!
+
+This will configure, build and run the `helloworld` application, you should see the following output:
```
```suggestion
This will configure, build and run the `helloworld` application.
You should see the following output:
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
 
-If you want to have more control you can also configure, build and run the application manually.
+The same can be done for `AArch64`.
+If you have built the application for `x86_64` before, run `make properclean` before the commands below.
```
```suggestion
If you have built the application for `x86_64` before, run `make properclean` before the commands below:
```

razvand commented on README.md:
````diff
@@ -1,63 +1,376 @@
 
-To configure it with the desired features:
+```console
+UK_DEFCONFIG=$(pwd)/.config.helloworld_qemu-arm64 make defconfig
````
```suggestion
make properclean
UK_DEFCONFIG=$(pwd)/.config.helloworld_qemu-arm64 make defconfig
```

razvand commented on README.md:
````diff
@@ -1,63 +1,376 @@
 ```
 
-Build the application:
+Now we will go into details about every step that we used above.
````
```suggestion
Information about every step is detailed below.
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
-Build the application:
+Now we will go into details about every step that we used above.
+
+## Requirements
```
```suggestion
### Requirements
```

razvand commented on README.md:
````diff
@@ -1,63 +1,376 @@
 ```
-$ make
+
+## Set Up
````
```suggestion
### Set Up
```

razvand commented on README.md:
````diff
@@ -1,63 +1,376 @@
+     10 directories, 7 files
+     ```
+
+## Configure
````
```suggestion
### Configure
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+
+Use the corresponding the configuration files (`config-...`), according to your choice of platform and architecture.
+
+### QEMU x86_64
```
```suggestion
#### QEMU x86_64
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+
+The `.config` file will be used in the build step.
+
+### QEMU AArch64
```
```suggestion
#### QEMU AArch64
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
-- If you built the application for `linuxu`:
+Similar to the x86_64 configuration, this results in the creation of the `.config` file that will be used in the build step.
+
+## Build
```
```suggestion
### Build
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+
+Typically, you would use `make properclean` to remove all build artifacts, but keep the configuration file.
+
+### QEMU x86_64
```
```suggestion
#### QEMU x86_64
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+Building uses as input the `.config` file from above, and results in a unikernel image as output.
+The unikernel output image, together with intermediary build files, are stored in the `build/` directory.
+
+### Clean Up
```
```suggestion
#### Clean Up
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+At the end of the build command, the `helloworld_qemu-x86_64` unikernel image is generated.
+This image is to be used in the run step.
+
+### QEMU AArch64
```
```suggestion
#### QEMU AArch64
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+Similarly to x86_64, at the end of the build command, the `helloworld_qemu-arm64` unikernel image is generated.
+This image is to be used in the run step.
+
+## Run
```
```suggestion
### Run
```

razvand commented on README.md:
```diff
@@ -1,63 +1,376 @@
+
+Run the resulting image using `qemu-system`.
+
+### QEMU x86_64
```
```suggestion
#### QEMU x86_64
```

razvand commented on README.md:
````diff
@@ -1,63 +1,376 @@
+Hello world!
+```
+
+### QEMU AArch64
````
```suggestion
#### QEMU AArch64
```

razvand reviewed (changes requested):
@StefanJum, see my comments.

StefanJum commented:
Check again @razvand

razvand commented on README.md:
```diff
@@ -1,63 +1,378 @@
+## Work with the Basic Build & Run Toolchain (Advanced)
+
+You can set up, configure, build and run the application from grounds up, without using the companion tool `kraft`.
+If you do not care about the reasoning behind the commands, you can just run them as listed below.
```
```suggestion

### Quick Setup (aka TLDR)

For a quick setup, run the commands below.
```

razvand reviewed (commented):
I made a suggestion.

razvand commented on README.md:
````diff
@@ -1,63 +1,378 @@
 
-To be able to run it, configure the application to run on the desired platform and architecture:
+```console
+curl --proto '=https' --tlsv1.2 -sSf https://get.kraftkit.sh | sh
````
Use `sudo sh`.

razvand commented on README.md:
````diff
@@ -1,63 +1,378 @@
 ```
-$ kraft configure -p PLATFORM -m ARCH
+
+Once installed, simple clone this repository and run `kraft build`:
````
```suggestion
Once installed, clone [this repository](https://github.com/unikraft/app-helloworld) and run `kraft build`:
```

razvand commented on README.md:
````diff
@@ -1,63 +1,378 @@
+
+```console
+git clone --depth=1 https://github.com/unikraft/app-hellworld.git
+cd app-helloworld;
````
```suggestion
cd helloworld
```

razvand commented on README.md:
````diff
@@ -1,63 +1,378 @@
+Once installed, simple clone this repository and run `kraft build`:
+
+```console
+git clone --depth=1 https://github.com/unikraft/app-hellworld.git
````
```suggestion
git clone https://github.com/unikraft/app-hellworld helloworld
```

StefanJum commented:
Done @razvand

razvand reviewed (changes requested):
Some other comments.

StefanJum commented:
Done

razvand commented on README.md:
````diff
@@ -1,63 +1,381 @@
+Once installed, clone [this repository](https://github.com/unikraft/app-helloworld) and run `kraft build`:
+
+```console
+git clone https://github.com/unikraft/app-hellworld helloworld
````
```suggestion
git clone https://github.com/unikraft/app-helloworld helloworld
```

razvand reviewed (approved):
Thanks, @StefanJum.

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

nderjung commented:
Thank you for incorporating the changes @StefanJum :)

razvand commented on README.md:
````diff
@@ -1,63 +1,381 @@
-$ kraft run
-Hello World!
+
+This will configure, build and run the `helloworld` application.
+You should see the following output:
+
+```text
+qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.vmx [bit 5]
+1: Set IPv4 address 172.44.0.2 mask 255.255.255.0 gw 172.44.0.1
+en1: Added
+en1: Interface is up
+Powered by
+o.   .o       _ _               __ _
+Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
+oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
+oOo oOO| | | | |   (| | | (_) |  _) :_
+ OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
+                  Atlas 0.13.1~5eb820bd
+Hello world!
 ```
````
I would remove this. It takes too much step for a TLDR.

razvand commented on README.md:
```diff
@@ -1,63 +1,381 @@
 
-If you want to have more control you can also configure, build and run the application manually.
+The same can be done for `AArch64`.
+If you have built the application for `x86_64` before, run `make properclean` before the commands below:
```
I wouldn't add this as part of TLDR. Just say "Run the commands below:"

razvand commented on README.md:
````diff
@@ -1,63 +1,381 @@
 ```
-$ make menuconfig
+
+Similar to the `x86_64` build, you will see the `helloworld` application running:
+
+```text
+qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.vmx [bit 5]
+Powered by
+o.   .o       _ _               __ _
+Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
+oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
+oOo oOO| | | | |   (| | | (_) |  _) :_
+ OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
+                  Atlas 0.13.1~5eb820bd
+Hello world!
 ```
````
I wouldn't add this as part of the TLDR. Keep it as simple as possible.

razvand reviewed (changes requested):
Some minor comments, @StefanJum

razvand commented on README.md:
```diff
@@ -1,63 +1,351 @@
-# Unikraft "hello world" Application
+# A Simple "Hello, world!" Application on Unikraft
```
```suggestion
# A "Hello, world!" Application on Unikraft
```

Avoid using simple, simply, easy, easily. It's never easy for newcomers, and it gives the impression that they're not up to the job.

razvand commented on README.md:
````diff
@@ -1,63 +1,351 @@
+/usr/bin/qemu-system-x86_64 -kernel build/helloworld_qemu-x86_64 -nographic
+```
+
+This will configure, build and run the `helloworld` application.
````
..., resulting in a `Hello world!` message being printed.

razvand commented on README.md:
````diff
@@ -1,63 +1,351 @@
+/usr/bin/qemu-system-aarch64 -kernel build/helloworld_qemu-arm64 -nographic -machine virt -cpu cortex-a57
+```
+
+Similar to the `x86_64` build, you will see the `helloworld` application running.
````
```suggestion
Similar to the `x86_64` build, this will result in a `Hello world!` message being printed.
```

razvand reviewed (changes requested):
Some other minor comments.

razvand reviewed (approved):
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
